src/main/java/org/jenkinsci/plugins/mesos/MesosSlave.java

```diff
@@ -27,32 +28,50 @@
   // Holds the current USI status for this agent.
   Optional<PodStatus> currentStatus = Optional.empty();
 
+  private final Boolean reusable;
+
+  private final MesosCloud cloud;
+
+  private final ExecutorService executorService;
+
+  private final String podId;
+
   public MesosSlave(
-      String name,
+      MesosCloud cloud,
+      String id,
       String nodeDescription,
       String labelString,
       List<? extends NodeProperty<?>> nodeProperties)
       throws Descriptor.FormException, IOException {
     super(
-        name,
-        nodeDescription,
-        null,
-        1,
-        null,
-        labelString,
-        new JNLPLauncher(),
-        null,
-        nodeProperties);
+        id, nodeDescription, null, 1, null, labelString, new JNLPLauncher(), null, nodeProperties);
+
+    // pass around the MesosApi connection via MesosCloud
+    this.cloud = cloud;
+    this.reusable = true;
+    this.executorService = Executors.newSingleThreadExecutor();
+    this.podId = id;
   }
 
   /**
    * Polls the agent until it is online. Note: This is a non-blocking call in contrast to the
    * blocking {@link AbstractCloudComputer#waitUntilOnline}.
    *
-   * @return This agent when it's online.
+   * @return The future agent that will come online.
    */
-  public CompletableFuture<MesosSlave> waitUntilOnlineAsync() {
-    throw new NotImplementedException();
+  public Future<MesosSlave> waitUntilOnlineAsync() {
+    return executorService.submit(
```

Comment:
I wonder if we should use Akka's [scheduler](https://doc.akka.io/japi/akka/current/index.html?akka/actor/package-summary.html) since that would be non-blocking. This call still block the thread.

```java
public Future<MesosSlave> waitUntilOnlineAsync() {
    var promise = new CompletableFuture<MesosSlave>();
    actorSystem.scheduleOnce(pollInterval, this::pollAgent(promise));
    return promise;
}

public void pollAgent(CompletableFuture<MesosSlave> promise) {
    if(this.toComputer().isOnline()) {
        promise.complete(this);
    } else {
        actorSystem.scheduleOnce(pollInterval, this::pollAgent(promise));
    }
}
```

Reply:
@meln1k actually has a nice solution with `Source.tick`.

Reply:
```scala
Source.tick(1.second, 2.second, NotUsed)
  .completionTimeout(5.minutes)
  .filter(_.isRunning)
  .toMat(Sink.head)(Keep.right)
  .run()
```

Reply:
Nice. It should be `isOnline` though.

Reply:
cool I will rebase on that when its ready

Reply:
Unfortunately not. Depending on how Awaitlity is implemented we are still blocking the thread. Akka's `Source.tick` won't block the thread nor eat up its resources. Furthermore, we get backpressure, ie if `isOnline/isRunning` takes longer than the tick interval Akka streams will just drop the next tick. I'm not sure that Awaitility is doing that.

Reply:
use akkas source.tick